crates/language/wasm/src/stdio.c

```diff
@@ -0,0 +1,310 @@
+            strncpy(temp_buf, str, precision);
+            temp_buf[precision] = '\0';
+            output_str = temp_buf;
+          } else {
+            output_str = str;
+          }
+          break;
+        }
+        case 'd':
+        case 'i': {
+          int value = va_arg(args, int);
+          int_to_str(value, temp_buf, 10, true, false);
+          break;
+        }
+        case 'u': {
+          unsigned int value = va_arg(args, unsigned int);
+          int_to_str(value, temp_buf, 10, false, false);
+          break;
+        }
+        case 'x': {
+          unsigned int value = va_arg(args, unsigned int);
+          int_to_str(value, temp_buf, 16, false, false);
+          break;
+        }
+        case 'X': {
+          unsigned int value = va_arg(args, unsigned int);
+          int_to_str(value, temp_buf, 16, false, true);
+          break;
+        }
+        case 'p': {
+          void *ptr = va_arg(args, void*);
+          ptr_to_str(ptr, temp_buf);
+          break;
+        }
+        case 'c': {
+          int c = va_arg(args, int);
+          temp_buf[0] = (char)c;
+          temp_buf[1] = '\0';
+          break;
+        }
+        case 'z': {
+          if (*(p + 1) == 'u') {
+            size_t value = va_arg(args, size_t);
+            int_to_str(value, temp_buf, 10, false, false);
+            p++;
+          } else {
+            temp_buf[0] = 'z';
+            temp_buf[1] = '\0';
+          }
+          break;
+        }
+        default:
+          temp_buf[0] = '%';
+          temp_buf[1] = *p;
+          temp_buf[2] = '\0';
+          break;
+      }
+
+      int str_len = strlen(output_str);
+      int formatted_len = (width > str_len) ? width : str_len;
+      total_chars += formatted_len;
+
+      if (pos < buffsz - 1) {
+        write_formatted_to_buffer(buffer, buffsz, &pos, output_str, width, &flags);
+      }
+
+    } else {
+      if (pos < buffsz - 1) buffer[pos++] = *p;
+      total_chars++;
+    }
+    p++;
+  }
+
+  if (buffsz > 0) buffer[pos < buffsz ? pos : buffsz - 1] = '\0';
+
+  return total_chars;
+}
+
+int snprintf(char *restrict buffer, size_t buffsz, const char *restrict format, ...) {
+  if (!buffer || buffsz == 0 || !format) return -1;
+
+  va_list args;
+  va_start(args, format);
+  int result = vsnprintf_impl(buffer, buffsz, format, args);
+  va_end(args);
+
+  return result;
+}
+
+int vsnprintf(char *restrict buffer, size_t buffsz, const char *restrict format, va_list vlist) {
+  return vsnprintf_impl(buffer, buffsz, format, vlist);
+}
+
+int fclose(FILE *stream) {
+  return 0;
+}
+
+FILE* fdopen(int fd, const char *mode) {
+  return 0;
+}
+
+int fputc(int c, FILE *stream) {
+  return c;
+}
+
+int fputs(const char *restrict str, FILE *restrict stream) {
+  return 0;
+}
+
+size_t fwrite(const void *restrict buffer, size_t size, size_t nmemb, FILE  *stream) {
+  return size * nmemb;
+}
+
+int fprintf(FILE *restrict stream, const char *restrict format, ...) {
+  return 0;
+}
```

Comment:
Should these `abort()`?

Reply:
No, if for some reason someone mistakenly calls `print_dot_graphs`, a no-op would be better than aborting, and I'd rather not litter all calls to these functions with `#ifdefs` just for one target triple.